File: LeetCode/LeetCode/OtherActivities/AlgorithmStudyPlan/Algorithm2/Problem33.cs
namespace LeetCode.OtherActivities.AlgorithmStudyPlan.Algorithm2
{
    public class Problem33
    {
        public int Search(int[] nums, int target)
        {
            int left = 0, mid;
            var right = nums.Length - 1;
            while (true)
            {
                mid = (right + left) / 2;
                if (nums[mid] == target)
                    return mid;

                if (left > right)
                    return -1;

                if (nums[left] <= nums[mid])
                {
                    if (nums[left] <= target && target <= nums[mid])
                    {
                        right = mid - 1;
                    }
                    else
                    {
                        left = mid + 1;
                    }
                }
                else
                {
                    if (nums[mid] <= target && target <= nums[right])
                    {
                        left = mid + 1;
                    }
                    else
                    {
                        right = mid - 1;
                    }
                }
            }
        }
    }
}

File: leetcode_rust/src/problems/problem_33.rs
pub fn search(nums: Vec<i32>, target: i32) -> i32 {
    let n = nums.len();

    let shift = super::helpers::get_shift_of_rotated_vector(n, &nums);

    let mut left = 0usize;
    let mut right = n - 1;

    while right - left > 1 {
        let mid = (right + left) / 2;
        let shifted_mid = (mid + shift) % n;
        let mid_el = nums[shifted_mid];
        if mid_el == target { return shifted_mid as i32; }
        if mid_el > target { right = mid; } else { left = mid; }
    }

    let shifted_right = (right + shift) % n;
    let shifted_left = (left + shift) % n;

    if nums[shifted_right] == target { shifted_right as i32 } else if nums[shifted_left] == target { shifted_left as i32 } else { -1 }
}
